qualtran/bloqs/generalized_qsp.py

```diff
@@ -220,23 +222,36 @@ def qsp_phase_factors(
     phi = np.zeros(n)
     lambd = 0
 
+    def safe_angle(x):
+        return 0 if np.isclose(x, 0) else np.angle(x)
+
     for d in reversed(range(n)):
         assert S.shape == (2, d + 1)
 
         a, b = S[:, d]
-        theta[d] = np.arctan2(np.abs(b), np.abs(a))
-        # \phi_d = arg(a / b)
-        # TODO this is numerically unstable for values of a and b around 0,
-        #      np.angle(0) = 0, but for near-zero values it produces the "correct angle",
-        #      even though we want it to be 0.
-        #      There should be a more numerically stable way to compute the relevant phis
-        #      so that the final QSP sequence is valid.
-        phi[d] = np.angle(a) - np.angle(b)
+        if only_real_coeffs:
+            theta[d] = np.arctan2(np.real(b), np.real(a))
+            phi[d] = 0
+        else:
+            theta[d] = np.arctan2(np.abs(b), np.abs(a))
+            # \phi_d = arg(a / b)
+            # TODO this is numerically unstable for values of a and b around 0,
+            #      np.angle(0) = 0, but for near-zero values it produces the "correct angle",
+            #      even though we want it to be 0.
+            #      There should be a more numerically stable way to compute the relevant phis
+            #      so that the final QSP sequence is valid.
+            phi[d] = 0 if np.isclose(np.abs(b), 0) else safe_angle(a) - safe_angle(b)
 
         if d == 0:
-            lambd = np.angle(b)
+            lambd = safe_angle(b)
         else:
             S = SU2RotationGate(theta[d], phi[d], 0).rotation_matrix @ S
+
+            # check if the gate was actually correct
+            aa, bb = S[0][0], S[1][d]
+            np.testing.assert_almost_equal(aa, 0, decimal=6)
+            np.testing.assert_almost_equal(bb, 0, decimal=6)
```

Comment:
@tanujkhattar I narrowed it down to this. These values must be 0, but are not iff the case fails (according to the unittests). I'll now investigate further why this is not working.